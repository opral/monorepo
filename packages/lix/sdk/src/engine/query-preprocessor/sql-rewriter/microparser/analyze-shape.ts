import {
	AS,
	AtName,
	ColonName,
	DollarName,
	DollarNumber,
	Dot,
	Ident,
	LIMIT,
	Num,
	QIdent,
	QMark,
	SQStr,
	type Token,
} from "../tokenizer.js";
import { findTableFactor, type TableFactorMatch } from "./table-factor.js";

export type PlaceholderToken = Token;

export type ValueLit =
	| { kind: "string"; value: string; token: Token }
	| { kind: "number"; value: number; token: Token };

export type Value = ValueLit | { kind: "placeholder"; token: PlaceholderToken };

export type ColumnRef = {
	alias?: string;
	column: string;
	startIx: number;
	endIx: number;
};

export type Filter = {
	lhs: ColumnRef;
	op: string;
	rhs: Value | { kind: "list"; values: Value[] } | null;
};

export type Limit =
	| { kind: "number"; value: number; token: Token }
	| { kind: "placeholder"; token: PlaceholderToken }
	| { kind: "none" };

export type Shape = {
	table: TableFactorMatch;
	filters: Filter[];
	limit: Limit;
	schemaKeys: Array<
		{ kind: "literal"; value: string } | { kind: "placeholder" }
	>;
	entityIds: Array<
		{ kind: "literal"; value: string } | { kind: "placeholder" }
	>;
	versionId:
		| { kind: "current" }
		| { kind: "literal"; value: string }
		| { kind: "placeholder"; token: PlaceholderToken }
		| { kind: "unknown" };
};

const placeholderTypes = new Set([
	QMark,
	ColonName,
	DollarName,
	DollarNumber,
	AtName,
]);

const isIdent = (token: Token | undefined): token is Token =>
	!!token && (token.tokenType === Ident || token.tokenType === QIdent);

const dequote = (image: string): string =>
	image.startsWith('"') && image.endsWith('"')
		? image.slice(1, -1).replace(/""/g, "")
		: image;

const normalize = (image: string): string => dequote(image).toLowerCase();

const unquoteString = (image: string): string =>
	image.slice(1, -1).replace(/''/g, "'");

const isSchemaColumn = (name: string) =>
	name === "schema_key" || name === "entity_id" || name === "version_id";

export function analyzeShape(tokens: Token[]): Shape | null {
	const table = findTableFactor(tokens, "internal_state_vtable");
	if (!table) return null;

	const filters: Filter[] = [];

	for (let i = 0; i < tokens.length; i++) {
		const token = tokens[i];

		// match qualified alias.column
		let alias: string | undefined;
		let columnToken: Token | undefined;

		if (
			isIdent(token) &&
			tokens[i + 1]?.tokenType === Dot &&
			isIdent(tokens[i + 2])
		) {
			alias = dequote(token.image);
			columnToken = tokens[i + 2];
		} else if (isIdent(token)) {
			columnToken = token;
		} else {
			continue;
		}

		if (!columnToken) {
			continue;
		}

		const columnName = normalize(columnToken.image);
		if (!isSchemaColumn(columnName)) continue;

		const result = parseFilter(tokens, i, alias, columnName);
		if (!result) {
			continue;
		}

		filters.push(result.filter);
		i = result.nextIndex;
	}

	const limit = readLimit(tokens);
	const schemaKeys = pluckValues(filters, "schema_key");
	const entityIds = pluckValues(filters, "entity_id");
	const versionId = determineVersionId(filters, tokens);

	return {
		table,
		filters,
		limit,
		schemaKeys,
		entityIds,
		versionId,
	};
}

function parseFilter(
	tokens: Token[],
	index: number,
	alias: string | undefined,
	column: string
): { filter: Filter; nextIndex: number } | null {
	let cursor = index + (alias ? 3 : 1); // skip alias, dot, column if qualified

	const opToken = tokens[cursor];
	if (!opToken) return null;

	const opImage = opToken.image.toUpperCase();
	if (opImage !== "=" && opImage !== "IN") return null;

	cursor += 1;

	if (opImage === "IN") {
		const rhs = parseInList(tokens, cursor);
		if (!rhs) return null;

		return {
			filter: {
				lhs: { alias, column, startIx: index, endIx: rhs.nextIndex },
				op: opImage,
				rhs: rhs.value,
			},
			nextIndex: rhs.nextIndex,
		};
	}

	const valueToken = tokens[cursor];
	if (!valueToken) return null;

	const rhs = toValue(valueToken);
	if (!rhs) return null;

	return {
		filter: {
			lhs: { alias, column, startIx: index, endIx: cursor },
			op: opImage,
			rhs,
		},
		nextIndex: cursor,
	};
}

function parseInList(
	tokens: Token[],
	index: number
): { value: { kind: "list"; values: Value[] }; nextIndex: number } | null {
	let cursor = index;
	if (tokens[cursor]?.image !== "(") return null;
	cursor += 1;

	const values: Value[] = [];

	while (cursor < tokens.length && tokens[cursor]?.image !== ")") {
		const token = tokens[cursor];
		if (!token) break;

		const value = toValue(token);
		if (!value) return null;
		values.push(value);

		cursor += 1;
		if (tokens[cursor]?.image === ",") {
			cursor += 1;
		}
	}

	if (cursor >= tokens.length || tokens[cursor]?.image !== ")") return null;

	const listValue: { kind: "list"; values: Value[] } = { kind: "list", values };
	return { value: listValue, nextIndex: cursor };
}

function toValue(token: Token): Value | null {
	if (token.tokenType === SQStr) {
		return { kind: "string", value: unquoteString(token.image), token };
	}
	if (token.tokenType === Num) {
		return { kind: "number", value: Number(token.image), token };
	}
	if (placeholderTypes.has(token.tokenType)) {
		return { kind: "placeholder", token };
	}
	return null;
}

function readLimit(tokens: Token[]): Limit {
	for (let i = 0; i < tokens.length - 1; i++) {
		const current = tokens[i];
		if (!current || current.tokenType !== LIMIT) continue;
		const next = tokens[i + 1];
		if (!next) break;

		if (next.tokenType === Num) {
			const value = Number(next.image);
			if (Number.isFinite(value)) {
				return { kind: "number", value, token: next };
			}
			return { kind: "none" };
		}

		if (next.tokenType === SQStr && next.image === "'1'") {
			return { kind: "number", value: 1, token: next };
		}

		if (placeholderTypes.has(next.tokenType)) {
			return { kind: "placeholder", token: next };
		}
	}

	return { kind: "none" };
}

function pluckValues(
	filters: Filter[],
	column: "schema_key" | "entity_id"
): Array<{ kind: "literal"; value: string } | { kind: "placeholder" }> {
	const results: Array<
		{ kind: "literal"; value: string } | { kind: "placeholder" }
	> = [];

	for (const filter of filters) {
		if (filter.lhs.column !== column) continue;
		const rhs = filter.rhs;
		if (!rhs) continue;

		if (rhs.kind === "list") {
			for (const value of rhs.values) {
				if (value.kind === "string") {
					results.push({ kind: "literal", value: value.value });
				} else if (value.kind === "placeholder") {
					results.push({ kind: "placeholder" });
				}
			}
		} else if (rhs.kind === "string") {
			results.push({ kind: "literal", value: rhs.value });
		} else if (rhs.kind === "placeholder") {
			results.push({ kind: "placeholder" });
		}
	}

	return results;
}

function determineVersionId(
	filters: Filter[],
	tokens: Token[]
): Shape["versionId"] {
	const versionFilter = filters.find(
		(filter) => filter.lhs.column === "version_id"
	);
	if (!versionFilter) return { kind: "unknown" };

	const rhs = versionFilter.rhs;
	if (!rhs) return { kind: "unknown" };

	if (rhs.kind === "string") {
		return { kind: "literal", value: rhs.value };
	}

	if (rhs.kind === "placeholder") {
		return { kind: "placeholder", token: rhs.token };
	}
	return { kind: "unknown" };
}
