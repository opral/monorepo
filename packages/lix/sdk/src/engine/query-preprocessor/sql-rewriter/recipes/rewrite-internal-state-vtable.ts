import type { Shape } from "../microparser/analyze-shape.js";

/**
 * Build a replacement subquery for `internal_state_vtable`.
 * Returns `null` when the query shape is outside the supported surface.
 */
export function rewriteInternalStateVtableQuery(shape: Shape): string | null {
	const schemaKey = pickSingleLiteral(shape.schemaKeys);
	const entityId = pickSingleLiteral(shape.entityIds);
	const versionId =
		shape.versionId.kind === "literal" ? shape.versionId.value : undefined;

	const limitOne = shape.limit.kind === "number" && shape.limit.value === 1;
	const literalOnly =
		!hasPlaceholder(shape.schemaKeys) && !hasPlaceholder(shape.entityIds);

	if (schemaKey && limitOne && literalOnly) {
		return buildFastPath({ schemaKey, entityId, versionId });
	}

	return buildWidePath({ schemaKeys: schemaKey ? [schemaKey] : [] });
}

interface FastPathOptions {
	schemaKey: string;
	entityId?: string;
	versionId?: string;
}

interface WidePathOptions {
	schemaKeys: string[];
}

const CACHE_SOURCE_TOKEN = "__CACHE_SOURCE__";

function buildFastPath(options: FastPathOptions): string {
	const filtersTxn = buildFilters("txn", options);
	const filtersUnt = buildFilters("u", options);
	const filtersCache = buildFilters("c", options);

	const txnSegment = stripIndent(`
    SELECT
      'T' || '~' || lix_encode_pk_part(txn.file_id) || '~' || lix_encode_pk_part(txn.entity_id) || '~' || lix_encode_pk_part(txn.version_id) AS _pk,
      txn.entity_id,
      txn.schema_key,
      txn.file_id,
      txn.plugin_key,
      json(txn.snapshot_content) AS snapshot_content,
      txn.schema_version,
      txn.version_id,
      txn.created_at,
      txn.created_at AS updated_at,
      NULL AS inherited_from_version_id,
      txn.id AS change_id,
      txn.untracked,
      'pending' AS commit_id,
      json(txn.metadata) AS metadata,
      ws_txn.writer_key
    FROM internal_transaction_state txn
    LEFT JOIN internal_state_writer ws_txn ON
      ws_txn.file_id = txn.file_id AND
      ws_txn.entity_id = txn.entity_id AND
      ws_txn.schema_key = txn.schema_key AND
      ws_txn.version_id = txn.version_id
    ${filtersTxn.length ? `WHERE ${filtersTxn.join("\n      AND ")}` : ""}
  `).trim();

	const untrackedBaseConditions = [
		"( (u.inheritance_delete_marker = 0 AND u.snapshot_content IS NOT NULL) OR\n      (u.inheritance_delete_marker = 1 AND u.snapshot_content IS NULL) )",
		`NOT EXISTS (
        SELECT 1 FROM internal_transaction_state t
        WHERE t.version_id = u.version_id
          AND t.file_id = u.file_id
          AND t.schema_key = u.schema_key
          AND t.entity_id = u.entity_id
      )`,
	];
	const untrackedSegment = stripIndent(`
    SELECT
      'U' || '~' || lix_encode_pk_part(u.file_id) || '~' || lix_encode_pk_part(u.entity_id) || '~' || lix_encode_pk_part(u.version_id) AS _pk,
      u.entity_id,
      u.schema_key,
      u.file_id,
      u.plugin_key,
      json(u.snapshot_content) AS snapshot_content,
      u.schema_version,
      u.version_id,
      u.created_at,
      u.updated_at,
      NULL AS inherited_from_version_id,
      'untracked' AS change_id,
      1 AS untracked,
      'untracked' AS commit_id,
      NULL AS metadata,
      ws_untracked.writer_key
    FROM internal_state_all_untracked u
    LEFT JOIN internal_state_writer ws_untracked ON
      ws_untracked.file_id = u.file_id AND
      ws_untracked.entity_id = u.entity_id AND
      ws_untracked.schema_key = u.schema_key AND
      ws_untracked.version_id = u.version_id
    WHERE ${[...untrackedBaseConditions, ...filtersUnt].join("\n      AND ")}
  `).trim();

	const cacheBaseConditions = [
		"( (c.inheritance_delete_marker = 0 AND c.snapshot_content IS NOT NULL) OR\n      (c.inheritance_delete_marker = 1 AND c.snapshot_content IS NULL) )",
		`NOT EXISTS (
        SELECT 1 FROM internal_transaction_state t
        WHERE t.version_id = c.version_id
          AND t.file_id = c.file_id
          AND t.schema_key = c.schema_key
          AND t.entity_id = c.entity_id
      )`,
		`NOT EXISTS (
        SELECT 1 FROM internal_state_all_untracked u
        WHERE u.version_id = c.version_id
          AND u.file_id = c.file_id
          AND u.schema_key = c.schema_key
          AND u.entity_id = c.entity_id
      )`,
	];

	const cacheSegment = stripIndent(`
    SELECT
      'C' || '~' || lix_encode_pk_part(c.file_id) || '~' || lix_encode_pk_part(c.entity_id) || '~' || lix_encode_pk_part(c.version_id) AS _pk,
      c.entity_id,
      c.schema_key,
      c.file_id,
      c.plugin_key,
      json(c.snapshot_content) AS snapshot_content,
      c.schema_version,
      c.version_id,
      c.created_at,
      c.updated_at,
      c.inherited_from_version_id,
      c.change_id,
      0 AS untracked,
      c.commit_id,
      ch.metadata AS metadata,
      ws_cache.writer_key
    FROM internal_state_cache c
    LEFT JOIN change ch ON ch.id = c.change_id
    LEFT JOIN internal_state_writer ws_cache ON
      ws_cache.file_id = c.file_id AND
      ws_cache.entity_id = c.entity_id AND
      ws_cache.schema_key = c.schema_key AND
      ws_cache.version_id = c.version_id
    WHERE ${[...cacheBaseConditions, ...filtersCache].join("\n      AND ")}
  `).trim();

	return stripIndent(`
    SELECT *
    FROM (
      ${txnSegment}

      UNION ALL

      ${untrackedSegment}

      UNION ALL

      ${cacheSegment}
    )
    LIMIT 1
  `);
}

function buildWidePath(options: WidePathOptions): string {
	const cacheSource = options.schemaKeys.length
		? `(${buildCacheRoutingSql(options.schemaKeys)})`
		: "internal_state_cache";

	const base = [buildVersionCtes(), buildResolvedSelectBody()].join("\n");
	return base.replaceAll(CACHE_SOURCE_TOKEN, cacheSource);
}

function buildFilters(alias: string, options: FastPathOptions): string[] {
	const filters: string[] = [
		`${alias}.schema_key = '${escapeLiteral(options.schemaKey)}'`,
	];
	if (options.entityId) {
		filters.push(`${alias}.entity_id = '${escapeLiteral(options.entityId)}'`);
	}
	if (options.versionId) {
		filters.push(`${alias}.version_id = '${escapeLiteral(options.versionId)}'`);
	}
	return filters;
}

function buildCacheRoutingSql(schemaKeys: string[]): string {
	return schemaKeys
		.map((key) =>
			stripIndent(`
        SELECT *
        FROM internal_state_cache
        WHERE schema_key = '${escapeLiteral(key)}'
      `)
		)
		.join("\nUNION ALL\n");
}

function buildVersionCtes(): string {
	return stripIndent(`
    WITH RECURSIVE
      version_descriptor_base AS (
        SELECT
          json_extract(isc_v.snapshot_content, '$.id') AS version_id,
          json_extract(isc_v.snapshot_content, '$.inherits_from_version_id') AS inherits_from_version_id
        FROM ${CACHE_SOURCE_TOKEN} isc_v
        WHERE isc_v.schema_key = 'lix_version_descriptor'
      ),
      version_inheritance(version_id, ancestor_version_id) AS (
        SELECT
          vdb.version_id,
          vdb.inherits_from_version_id
        FROM version_descriptor_base vdb
        WHERE vdb.inherits_from_version_id IS NOT NULL

        UNION

        SELECT
          vir.version_id,
          vdb.inherits_from_version_id
        FROM version_inheritance vir
        JOIN version_descriptor_base vdb ON vdb.version_id = vir.ancestor_version_id
        WHERE vdb.inherits_from_version_id IS NOT NULL
      ),
      version_parent AS (
        SELECT
          vdb.version_id,
          vdb.inherits_from_version_id AS parent_version_id
        FROM version_descriptor_base vdb
        WHERE vdb.inherits_from_version_id IS NOT NULL
      )
  `);
}

function buildResolvedSelectBody(): string {
	const segments = [
		buildTransactionSegment(),
		buildUntrackedSegment(),
		buildCachedSegment(),
		buildInheritedCacheSegment(),
		buildInheritedUntrackedSegment(),
		buildInheritedTransactionSegment(),
	];

	return [
		"      SELECT * FROM (",
		segments.join("\n\n\t\tUNION ALL\n\n"),
		"\t\t)",
	].join("\n");
}

function buildTransactionSegment(): string {
	return stripIndent(`
      -- 1. Transaction state (highest priority) - pending changes
      SELECT 
        'T' || '~' || lix_encode_pk_part(txn.file_id) || '~' || lix_encode_pk_part(txn.entity_id) || '~' || lix_encode_pk_part(txn.version_id) AS _pk,
        txn.entity_id,
        txn.schema_key,
        txn.file_id,
        txn.plugin_key,
        json(txn.snapshot_content) AS snapshot_content,
        txn.schema_version,
        txn.version_id,
        txn.created_at,
        txn.created_at AS updated_at,
        NULL AS inherited_from_version_id,
        txn.id AS change_id,
        txn.untracked,
        'pending' AS commit_id,
        json(txn.metadata) AS metadata,
        ws_txn.writer_key
      FROM internal_transaction_state txn
      LEFT JOIN internal_state_writer ws_txn ON
        ws_txn.file_id = txn.file_id AND
        ws_txn.entity_id = txn.entity_id AND
        ws_txn.schema_key = txn.schema_key AND
        ws_txn.version_id = txn.version_id
  `);
}

function buildUntrackedSegment(): string {
	return stripIndent(`
      -- 2. Untracked state (second priority) - only if no transaction exists
      SELECT 
        'U' || '~' || lix_encode_pk_part(u.file_id) || '~' || lix_encode_pk_part(u.entity_id) || '~' || lix_encode_pk_part(u.version_id) AS _pk,
        u.entity_id,
        u.schema_key,
        u.file_id,
        u.plugin_key,
        json(u.snapshot_content) AS snapshot_content,
        u.schema_version,
        u.version_id,
        u.created_at,
        u.updated_at,
        NULL AS inherited_from_version_id,
        'untracked' AS change_id,
        1 AS untracked,
        'untracked' AS commit_id,
        NULL AS metadata,
        ws_untracked.writer_key
      FROM internal_state_all_untracked u
      LEFT JOIN internal_state_writer ws_untracked ON
        ws_untracked.file_id = u.file_id AND
        ws_untracked.entity_id = u.entity_id AND
        ws_untracked.schema_key = u.schema_key AND
        ws_untracked.version_id = u.version_id
      WHERE (
        (u.inheritance_delete_marker = 0 AND u.snapshot_content IS NOT NULL) OR
        (u.inheritance_delete_marker = 1 AND u.snapshot_content IS NULL)
      )
        AND NOT EXISTS (
          SELECT 1 FROM internal_transaction_state t
          WHERE t.version_id = u.version_id
            AND t.file_id = u.file_id
            AND t.schema_key = u.schema_key
            AND t.entity_id = u.entity_id
        )
  `);
}

function buildCachedSegment(): string {
	return stripIndent(`
      -- 3. Tracked state from cache (third priority) - only if no transaction or untracked exists
      SELECT 
        'C' || '~' || lix_encode_pk_part(c.file_id) || '~' || lix_encode_pk_part(c.entity_id) || '~' || lix_encode_pk_part(c.version_id) AS _pk,
        c.entity_id,
        c.schema_key,
        c.file_id,
        c.plugin_key,
        json(c.snapshot_content) AS snapshot_content,
        c.schema_version,
        c.version_id,
        c.created_at,
        c.updated_at,
        c.inherited_from_version_id,
        c.change_id,
        0 AS untracked,
        c.commit_id,
        ch.metadata AS metadata,
        ws_cache.writer_key
      FROM ${CACHE_SOURCE_TOKEN} c
      LEFT JOIN change ch ON ch.id = c.change_id
      LEFT JOIN internal_state_writer ws_cache ON
        ws_cache.file_id = c.file_id AND
        ws_cache.entity_id = c.entity_id AND
        ws_cache.schema_key = c.schema_key AND
        ws_cache.version_id = c.version_id
      WHERE (
        (c.inheritance_delete_marker = 0 AND c.snapshot_content IS NOT NULL) OR
        (c.inheritance_delete_marker = 1 AND c.snapshot_content IS NULL)
      )
        AND NOT EXISTS (
          SELECT 1 FROM internal_transaction_state t
          WHERE t.version_id = c.version_id
            AND t.file_id = c.file_id
            AND t.schema_key = c.schema_key
            AND t.entity_id = c.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM internal_state_all_untracked u
          WHERE u.version_id = c.version_id
            AND u.file_id = c.file_id
            AND u.schema_key = c.schema_key
            AND u.entity_id = c.entity_id
        )
  `);
}

function buildInheritedCacheSegment(): string {
	return stripIndent(`
      -- 4. Inherited tracked state (fourth priority)
      SELECT 
        'CI' || '~' || lix_encode_pk_part(isc.file_id) || '~' || lix_encode_pk_part(isc.entity_id) || '~' || lix_encode_pk_part(vi.version_id) AS _pk,
        isc.entity_id,
        isc.schema_key,
        isc.file_id,
        isc.plugin_key,
        json(isc.snapshot_content) AS snapshot_content,
        isc.schema_version,
        vi.version_id,
        isc.created_at,
        isc.updated_at,
        isc.version_id AS inherited_from_version_id,
        isc.change_id,
        0 AS untracked,
        isc.commit_id,
        ch.metadata AS metadata,
        COALESCE(ws_child.writer_key, ws_parent.writer_key) AS writer_key
      FROM version_inheritance vi
      JOIN ${CACHE_SOURCE_TOKEN} isc ON isc.version_id = vi.ancestor_version_id
      LEFT JOIN change ch ON ch.id = isc.change_id
      LEFT JOIN internal_state_writer ws_child ON
        ws_child.file_id = isc.file_id AND
        ws_child.entity_id = isc.entity_id AND
        ws_child.schema_key = isc.schema_key AND
        ws_child.version_id = vi.version_id
      LEFT JOIN internal_state_writer ws_parent ON
        ws_parent.file_id = isc.file_id AND
        ws_parent.entity_id = isc.entity_id AND
        ws_parent.schema_key = isc.schema_key AND
        ws_parent.version_id = isc.version_id
      WHERE isc.inheritance_delete_marker = 0
        AND isc.snapshot_content IS NOT NULL
        AND NOT EXISTS (
          SELECT 1 FROM internal_transaction_state t
          WHERE t.version_id = vi.version_id
            AND t.file_id = isc.file_id
            AND t.schema_key = isc.schema_key
            AND t.entity_id = isc.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM ${CACHE_SOURCE_TOKEN} child_isc
          WHERE child_isc.version_id = vi.version_id
            AND child_isc.file_id = isc.file_id
            AND child_isc.schema_key = isc.schema_key
            AND child_isc.entity_id = isc.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM internal_state_all_untracked child_unt
          WHERE child_unt.version_id = vi.version_id
            AND child_unt.file_id = isc.file_id
            AND child_unt.schema_key = isc.schema_key
            AND child_unt.entity_id = isc.entity_id
        )
  `);
}

function buildInheritedUntrackedSegment(): string {
	return stripIndent(`
      -- 5. Inherited untracked state (lowest priority)
      SELECT 
        'UI' || '~' || lix_encode_pk_part(unt.file_id) || '~' || lix_encode_pk_part(unt.entity_id) || '~' || lix_encode_pk_part(vi.version_id) AS _pk,
        unt.entity_id,
        unt.schema_key,
        unt.file_id,
        unt.plugin_key,
        json(unt.snapshot_content) AS snapshot_content,
        unt.schema_version,
        vi.version_id,
        unt.created_at,
        unt.updated_at,
        unt.version_id AS inherited_from_version_id,
        'untracked' AS change_id,
        1 AS untracked,
        'untracked' AS commit_id,
        NULL AS metadata,
        COALESCE(ws_child.writer_key, ws_parent.writer_key) AS writer_key
      FROM version_inheritance vi
      JOIN internal_state_all_untracked unt ON unt.version_id = vi.ancestor_version_id
      LEFT JOIN internal_state_writer ws_child ON
        ws_child.file_id = unt.file_id AND
        ws_child.entity_id = unt.entity_id AND
        ws_child.schema_key = unt.schema_key AND
        ws_child.version_id = vi.version_id
      LEFT JOIN internal_state_writer ws_parent ON
        ws_parent.file_id = unt.file_id AND
        ws_parent.entity_id = unt.entity_id AND
        ws_parent.schema_key = unt.schema_key AND
        ws_parent.version_id = unt.version_id
      WHERE unt.inheritance_delete_marker = 0
        AND unt.snapshot_content IS NOT NULL
        AND NOT EXISTS (
          SELECT 1 FROM internal_transaction_state t
          WHERE t.version_id = vi.version_id
            AND t.file_id = unt.file_id
            AND t.schema_key = unt.schema_key
            AND t.entity_id = unt.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM ${CACHE_SOURCE_TOKEN} child_isc
          WHERE child_isc.version_id = vi.version_id
            AND child_isc.file_id = unt.file_id
            AND child_isc.schema_key = unt.schema_key
            AND child_isc.entity_id = unt.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM internal_state_all_untracked child_unt
          WHERE child_unt.version_id = vi.version_id
            AND child_unt.file_id = unt.file_id
            AND child_unt.schema_key = unt.schema_key
            AND child_unt.entity_id = unt.entity_id
        )
  `);
}

function buildInheritedTransactionSegment(): string {
	return stripIndent(`
      -- 6. Inherited transaction state
      SELECT 
        'TI' || '~' || lix_encode_pk_part(txn.file_id) || '~' || lix_encode_pk_part(txn.entity_id) || '~' || lix_encode_pk_part(vi.version_id) AS _pk,
        txn.entity_id,
        txn.schema_key,
        txn.file_id,
        txn.plugin_key,
        json(txn.snapshot_content) AS snapshot_content,
        txn.schema_version,
        vi.version_id,
        txn.created_at,
        txn.created_at AS updated_at,
        vi.parent_version_id AS inherited_from_version_id,
        txn.id AS change_id,
        txn.untracked,
        'pending' AS commit_id,
        json(txn.metadata) AS metadata,
        COALESCE(ws_child.writer_key, ws_parent.writer_key) AS writer_key
      FROM version_parent vi
      JOIN internal_transaction_state txn ON txn.version_id = vi.parent_version_id
      LEFT JOIN internal_state_writer ws_child ON
        ws_child.file_id = txn.file_id AND
        ws_child.entity_id = txn.entity_id AND
        ws_child.schema_key = txn.schema_key AND
        ws_child.version_id = vi.version_id
      LEFT JOIN internal_state_writer ws_parent ON
        ws_parent.file_id = txn.file_id AND
        ws_parent.entity_id = txn.entity_id AND
        ws_parent.schema_key = txn.schema_key AND
        ws_parent.version_id = vi.parent_version_id
      WHERE vi.parent_version_id IS NOT NULL
        AND txn.snapshot_content IS NOT NULL
        AND NOT EXISTS (
          SELECT 1 FROM internal_transaction_state child_txn
          WHERE child_txn.version_id = vi.version_id
            AND child_txn.file_id = txn.file_id
            AND child_txn.schema_key = txn.schema_key
            AND child_txn.entity_id = txn.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM ${CACHE_SOURCE_TOKEN} child_isc
          WHERE child_isc.version_id = vi.version_id
            AND child_isc.file_id = txn.file_id
            AND child_isc.schema_key = txn.schema_key
            AND child_isc.entity_id = txn.entity_id
        )
        AND NOT EXISTS (
          SELECT 1 FROM internal_state_all_untracked child_unt
          WHERE child_unt.version_id = vi.version_id
            AND child_unt.file_id = txn.file_id
            AND child_unt.schema_key = txn.schema_key
            AND child_unt.entity_id = txn.entity_id
        )
  `);
}

type LiteralOrPlaceholder =
	| { kind: "literal"; value: string }
	| { kind: "placeholder" };

function pickSingleLiteral(values: LiteralOrPlaceholder[]): string | undefined {
	const literals = values.filter(
		(entry): entry is { kind: "literal"; value: string } =>
			entry.kind === "literal"
	);
	if (literals.length !== 1) return undefined;
	return literals[0]!.value;
}

function hasPlaceholder(values: LiteralOrPlaceholder[]): boolean {
	return values.some((entry) => entry.kind === "placeholder");
}

function stripIndent(value: string): string {
	const trimmed = value.replace(/^\n+/, "").replace(/\n+$/, "");
	const lines = trimmed.split("\n");
	const indents = lines
		.filter((line) => line.trim().length > 0)
		.map((line) => line.match(/^\s*/)?.[0]?.length ?? 0);
	const minIndent = indents.length > 0 ? Math.min(...indents) : 0;
	return lines.map((line) => line.slice(minIndent)).join("\n");
}

function escapeLiteral(value: string): string {
	return value.replace(/'/g, "''");
}
