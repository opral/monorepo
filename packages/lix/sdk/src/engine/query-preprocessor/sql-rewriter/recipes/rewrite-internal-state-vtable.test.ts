import { expect, test } from "vitest";
import { tokenize } from "../tokenizer.js";
import { analyzeShape } from "../microparser/analyze-shape.js";
import { rewriteInternalStateVtableQuery } from "./rewrite-internal-state-vtable.js";
import { openLix } from "../../../../lix/open-lix.js";

test("rewrites a simple internal_state_vtable select", () => {
	const sql = `SELECT * FROM internal_state_vtable;`;
	const tokens = tokenize(sql);
	const shape = analyzeShape(tokens);
	expect(shape).not.toBeNull();

	const rewritten = rewriteInternalStateVtableQuery(shape!);
	expect(rewritten).toBeTruthy();
	expect(rewritten).toContain("WITH RECURSIVE");
	expect(rewritten).toContain("internal_state_cache");
});

test("passes literal schema key into cache routing", () => {
	const sql = `SELECT * FROM internal_state_vtable v WHERE v.schema_key = 'lix_key_value';`;
	const tokens = tokenize(sql);
	const shape = analyzeShape(tokens);

	expect(shape).not.toBeNull();

	const rewritten = rewriteInternalStateVtableQuery(shape!);
	expect(rewritten).toBeTruthy();
	expect(rewritten).toContain("lix_key_value");
});

test("uses fast path for limit 1 queries", async () => {
	const sql = `SELECT * FROM internal_state_vtable v WHERE v.schema_key = 'lix_key_value' LIMIT 1;`;
	const tokens = tokenize(sql);
	const shape = analyzeShape(tokens);
	expect(shape).not.toBeNull();

	const rewritten = rewriteInternalStateVtableQuery(shape!);
	expect(rewritten).toBeTruthy();
	expect(rewritten).toContain("LIMIT 1");
	expect(rewritten).toContain("internal_transaction_state");
	expect(rewritten).not.toContain("WITH RECURSIVE");

	const rewrittenQuery = `SELECT * FROM (${rewritten}) AS v;`;
	const plan = await explainQueryPlan(rewrittenQuery);

	expect(plan).not.toContain("internal_state_vtable");
	expect(plan).toMatchSnapshot();
});

async function explainQueryPlan(sql: string): Promise<string> {
	const lix = await openLix({});
	const sanitized = sql.trim().replace(/;$/, "");
	try {
		const { rows } = lix.engine!.executeSync({
			sql: `EXPLAIN QUERY PLAN ${sanitized}`,
			parameters: [],
		});
		return rows
			.map((row) =>
				Object.values(row)
					.map((value) =>
						value === null || value === undefined ? "" : String(value)
					)
					.join("|")
			)
			.join("\n");
	} finally {
		await lix.close();
	}
}
