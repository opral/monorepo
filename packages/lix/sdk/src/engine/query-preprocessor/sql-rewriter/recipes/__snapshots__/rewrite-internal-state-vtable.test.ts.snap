// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uses fast path for limit 1 queries 1`] = `
"2|0|0|CO-ROUTINE (subquery-6)
3|2|0|COMPOUND QUERY
4|3|0|LEFT-MOST SUBQUERY
7|4|216|SCAN txn
11|4|45|SEARCH ws_txn USING PRIMARY KEY (file_id=? AND version_id=? AND entity_id=? AND schema_key=?) LEFT-JOIN
53|3|0|UNION ALL
56|53|216|SCAN u
69|53|0|CORRELATED SCALAR SUBQUERY 2
73|69|45|SEARCH t USING COVERING INDEX sqlite_autoindex_internal_transaction_state_2 (entity_id=? AND file_id=? AND schema_key=? AND version_id=?)
85|53|45|SEARCH ws_untracked USING PRIMARY KEY (file_id=? AND version_id=? AND entity_id=? AND schema_key=?) LEFT-JOIN
126|3|0|UNION ALL
129|126|0|MATERIALIZE change
131|129|0|COMPOUND QUERY
132|131|0|LEFT-MOST SUBQUERY
136|132|216|SCAN c
138|132|47|SEARCH s USING INDEX sqlite_autoindex_internal_snapshot_1 (id=?) LEFT-JOIN
163|131|0|UNION ALL
165|163|216|SCAN t
186|126|99|SCAN c VIRTUAL TABLE INDEX 0:schema_key
200|126|0|CORRELATED SCALAR SUBQUERY 4
204|200|45|SEARCH t USING COVERING INDEX sqlite_autoindex_internal_transaction_state_2 (entity_id=? AND file_id=? AND schema_key=? AND version_id=?)
217|126|0|CORRELATED SCALAR SUBQUERY 5
221|217|45|SEARCH u USING COVERING INDEX sqlite_autoindex_internal_state_all_untracked_1 (entity_id=? AND schema_key=? AND file_id=? AND version_id=?)
251|126|53|SEARCH ch USING AUTOMATIC COVERING INDEX (id=?) LEFT-JOIN
258|126|45|SEARCH ws_cache USING PRIMARY KEY (file_id=? AND version_id=? AND entity_id=? AND schema_key=?) LEFT-JOIN
305|0|207|SCAN (subquery-6)"
`;
